src/hlstrafe.cpp

```diff
@@ -1214,15 +1214,29 @@ namespace HLStrafe
 		}
 	}
 
-	PositionType GetPositionType(PlayerData& player, TraceFunc traceFunc)
+	PositionType GetPositionType(PlayerData& player, TraceFunc traceFunc, PointContentsFunc pointContentsFunc)
 	{
-		// TODO: Check water. If we're under water, return here.
+		// Check water. If we're under water, return here.
+		const int CONTENTS_WATER = -3;
+		const int CONTENTS_TRANSLUCENT = -15;
+
+		float point[3];
+		VecCopy<float, 3>(player.Origin, point);
+
+		// Pick a spot just above the players feet.
+		point[0] += (player.Ducking ? (VEC_DUCK_HULL_MIN[0] + VEC_DUCK_HULL_MAX[0]) : (VEC_HULL_MIN[0] + VEC_HULL_MAX[0])) * 0.5f;
+		point[1] += (player.Ducking ? (VEC_DUCK_HULL_MIN[1] + VEC_DUCK_HULL_MAX[1]) : (VEC_HULL_MIN[1] + VEC_HULL_MAX[1])) * 0.5f;
+		point[2] += ((player.Ducking ? VEC_DUCK_HULL_MIN[2] : VEC_HULL_MIN[2]) + 1);
+
+		int contents = pointContentsFunc(point);
+
+		if (contents <= CONTENTS_WATER && contents > CONTENTS_TRANSLUCENT)
```

Comment:
You still need the second point check. Right now this will return water even if it's very shallow and the player isn't swimming.